src/extension/handler/utils.ts

```diff
@@ -0,0 +1,139 @@
+const DEFAULT_START_FILE = 'README.md'
+
+/**
+ * Get the project directory from the settings object.
+ *
+ * If none is set up, the directory will be a temporary directory.
+ *
+ * If the set up directory doesn't exist, it will ask whether to create it and if not
+ * we return `null` meaning we should cancel the checkout operation.
+ *
+ * @returns `vscode.Uri` with target directory to check out the project to
+ *          `null` if the user doesn't want to create the directory, here we should cancel the operation
+ */
+export async function getProjectDir () {
+    const projectDirPath: string | undefined = config.get('projectDir')
+
+    if (!projectDirPath) {
+        return Uri.parse((await dir()).path)
+    }
+
+    const projectDir = Uri.parse(projectDirPath)
+    const isExisting = await workspace.fs.stat(projectDir)
+        .then(() => true, () => false)
+    if (isExisting) {
+        return projectDir
+    }
+
+    const createDir = (await window.showInformationMessage(
+        `A project directory (${projectDir}) was set up but doesn't exist. ` +
+        'Do you want to create it?',
+        'Yes', 'No'
+    )) === 'Yes'
+
+    if (!createDir) {
+        return null
+    }
+
+    await workspace.fs.createDirectory(projectDir)
+    return projectDir
+}
+
+/**
+ * Create the name of the target directory to checkout a project to
+ * @param targetDir Uri of the base directory (received by calling `getProjectDir`)
+ * @param suggestedName name of the directory to check the project into, e.g. "org/projectName"
+ * @param index index which increases if directory name exists (e.g. "foobar_1")
+ * @returns a string with the name of the target directory
+ */
+export async function getTargetDirName (targetDir: Uri, suggestedName: string, index = 0): Promise<string> {
+    /**
+     * for now let's expect a suggested name mimicking the format "<org>/<project>"
+     */
+    if (suggestedName.split('/').length !== 2) {
+        throw new Error(`Invalid project directory suggestion: ${suggestedName}`)
+    }
+
+    /**
+     * create org directory
+     */
+    const [orgName] = suggestedName.split('/')
+    const orgDir = Uri.joinPath(targetDir, orgName)
+    const isOrgDirExisting = await workspace.fs.stat(orgDir).then(() => true, () => false)
+    if (!isOrgDirExisting) {
+        await workspace.fs.createDirectory(orgDir)
+    }
+
+    const amendedSuggestedName = !index ? suggestedName: `${suggestedName}_${index}`
+    const fullTargetDir = Uri.joinPath(targetDir, amendedSuggestedName)
+    const isExisting = await workspace.fs.stat(fullTargetDir).then(() => true, () => false)
+    if (isExisting) {
+        return getTargetDirName(targetDir, suggestedName, ++index)
+    }
+
+    return amendedSuggestedName
+}
+
+export async function waitForProjectCheckout (
+    targetDir: string,
+    repository: string,
+    cb: (success: boolean) => void
+) {
+    const targetDirUri = Uri.parse(targetDir)
+    const t = setTimeout(() => {
```

Comment:
Let's do at least 60s by default, perphaps even more. 10 seconds is not enough for even fast/reliable internet. Example: clone into `microsoft/vscode`.

Reply:
@mxsdev any suggestion how to handle it differently. I could imagine having a _"This takes longer than expected"_ notification and allow users to cancel it, rather than having a hard timeout

> Let's do at least 60s by default

That's [the current behavior](https://github.com/stateful/vscode-runme/blob/7f02e41401bd539311144e5d0d32a988081b1100/package.json#L252).

> 10 seconds is not enough for even fast/reliable internet. Example: clone into microsoft/vscode.

10 seconds here is the minimum timeout, meaning that if someone sets it to 0s it will at least wait for 10s.

Reply:
Sorry, my bad. Misunderstood Max's command. Please ignore.

Reply:
@christian-bromann 

> 10 seconds here is the minimum timeout

Oh, my bad! I saw ten seconds and my brain rolled with it. In that case, if we're sticking with polling, then I think the current solution is fine, since the users can see the terminal output anyway.